src/math/bvh.js

```diff
@@ -0,0 +1,456 @@
+        this.bvhNode[leftChildIdx].leftFirst = node.leftFirst;
+        this.bvhNode[leftChildIdx].triCount = leftCount;
+        this.bvhNode[rightChildIdx].leftFirst = i;
+        this.bvhNode[rightChildIdx].triCount = node.triCount - leftCount;
+        node.leftFirst = leftChildIdx;
+        node.triCount = 0;
+        this.updateNodeBounds(leftChildIdx);
+        this.updateNodeBounds(rightChildIdx);
+
+        // Recurse into each of the child nodes
+        this.Subdivide(leftChildIdx);
+        this.Subdivide(rightChildIdx);
+    }
+
+    /**
+     * Builds the BVH
+     *
+     * @ignore
+     */
+    buildBVH() {
+        const { triangles } = this;
+        const N = triangles.length;
+        this.triIdx = Array.from({ length: N }, (x, i) => i);
+        this.bvhNode = Array.apply(null, Array(N * 2)).map(function () {
+            return new BVHNode();
+        });
+        const rootNodeIdx = 0;
+        this.nodesUsed = 1;
+
+        for (let i = 0; i < N; i++) {
+            triangles[i].centroid = new Vec3();
+            triangles[i].centroid.add(triangles[i].vertex0);
+            triangles[i].centroid.add(triangles[i].vertex1);
+            triangles[i].centroid.add(triangles[i].vertex2);
+            triangles[i].centroid.mulScalar(1 / 3);
+        }
+
+        const root = this.bvhNode[0];
+        root.leftFirst = 0;
+        root.triCount = N;
+        this.updateNodeBounds(rootNodeIdx);
+        this.Subdivide(rootNodeIdx);
+    }
+
+    /**
+     * Refits the BVH
+     *
+     * @param {Tri[]} triangles - The triangles to refit the BVH on
+     * @ignore
+     */
+    refitBVH(triangles) {
+        this.triangles = triangles;
+        for (let i = this.nodesUsed - 1; i >= 0; i--) {
+            if (i !== 1) {
+                const node = this.bvhNode[i];
+                if (node.isLeaf()) {
+                    // adjust bounds to contained triangles for leaf nodes
+                    this.updateNodeBounds(i);
+                    continue;
+                }
+                // adjust boudns to child node bounds in interior nodes
+                const leftChild = this.bvhNode[node.leftFirst];
+                const rightChild = this.bvhNode[node.leftFirst + 1];
+                vec3Min(leftChild.aabbMin, rightChild.aabbMin, node.aabbMin);
+                vec3Max(leftChild.aabbMax, rightChild.aabbMax, node.aabbMax);
+            }
+        }
+    }
+
+    /**
+     * Calculates the shortest distance between the origin of a ray and its intersection with the mesh
+     *
+     * @param {Ray} ray - The ray to be traced
+     * @param {number} nodeIdx - The index of the node to begin the raycast at (default is zero: the root)
+     * @ignore
+     */
+    intersectBVH(ray, nodeIdx = 0) {
+        ray.rDx = 1 / ray.direction.x;
+        ray.rDy = 1 / ray.direction.y;
+        ray.rDz = 1 / ray.direction.z;
+
+        let node = this.bvhNode[nodeIdx];
+        const stack = [];
```

Comment:
have stack as a property of this class
- created that in constructor as `this.stack = [];`
- to clear it, do `this.stack.lenght = 0;`  // this keeps it allocated, and avoids new array allocation

Reply:
Done